### PR TITLE
Java 9 support (WIP)

### PR DESCRIPTION
**Pull request by dgelessus**
Lots of minor changes. Now the source can be compiled with JDK 6, and more importantly with JDK 9 with Java 6 language level (see #294).

### CODE REVIEW

**Comment by dgelessus** [comment]
And just pushed some more Java 9 compatibility stuff, which all works on older Javas too of course. Still not able to run the launcher on Java 9 though, because the hack to add things to the classpath at runtime doesn't work anymore. I haven't found a good Java 9-compatible way of doing that yet.

**Comment by jamierocks** [comment]
Supporting SE 6 just encourages people to use it. I also see no reason for ATL to support it now, especially after its EOL.

**Comment by RyanTheAllmighty** [comment]
AFAIK 6 is still required for our Mac app

**Comment by jamierocks** [comment]
Do you have stats on people that use Mac and SE 6?

**Comment by RyanTheAllmighty** [comment]
Everyone who uses the Mac application uses Java 6 because of the way it's packaged. So it's not an option to remove Java 6 compatibility without first porting over the changes made in 3.3.0.0 branch to move it away from Java 6

**Comment by dgelessus** [comment]
FWIW, I would personally prefer to keep Java 6 support, especially since this version of the launcher won't be updated anymore once NEXT is out. The differences between 6 and 7 are negligible, and dropping Java 6 support this late just to get the new syntax seems a bit unnecessary IMHO. (I'd be more in favor of requiring Java 7 if NEXT wasn't already in development.)

And although the number of users who actually need Java 6 is small, there are still lots of people using the old Mac app. If we were to drop Java 6 support, we'd need to figure out how to update the entire app bundle to the new version, not just the launcher jar as we're doing at the moment.

**Comment by dgelessus** [comment]
And pushed some more fixes. Also merged ATLauncher/feature/302-update-authlib into my develop branch already, because that change is sort of necessary for Java 9. The launcher now seems to run fine, installing instances works, and launching instances is fine as long as the Minecraft Java is not Java 9. (LaunchWrapper also uses the `URLClassLoader` hack, so it fails right at the beginning.)

I noticed that when I run the launcher with Java 9 and have the Minecraft Java set to Java 8, Forge picks up jars from my JDK 9 installation somehow and tries to load them as mods (and fails, obviously). This doesn't affect the game as Forge handles the error correctly, but I'll look into that and see why it's happening.

The code also works fine on older Javas, so this should be okay to merge now.

**Comment by dgelessus** [comment]
Ah, I see. [This](https://github.com/dgelessus/ATLauncher/blob/b85f5039b6dcb6580cdcf552334a18bdad2f3ac4/src/main/java/com/atlauncher/mclauncher/MCLauncher.java#L196) adds the entire classpath of the launcher Java to the Minecraft classpath. Is that even necessary? The `-cp` option should extend the classpath, and not overwrite it entirely, so there should be no need to add the standard classpath manually. I tried leaving out the launcher Java classpath and everything seems to be working fine, so I'll push that as well.

File: src/main/java/com/atlauncher/mclauncher/MCLauncher.java [hunk 1]
```diff
@@ -192,8 +189,19 @@ public static Process launch(Account account, Instance instance, LoginResponse r
         }
 
         arguments.add("-Djava.library.path=" + instance.getNativesDirectory().getAbsolutePath());
-        arguments.add("-cp");
-        arguments.add(System.getProperty("java.class.path") + cpb.toString());
+        
+        if (!cp.isEmpty()) {
```
Comment (jamierocks) [inline src/main/java/com/atlauncher/mclauncher/MCLauncher.java]: Its not ever going to be empty.

**Comment by jamierocks** [comment]
You need to take into account that ATLauncher's classpath doesn't just contain Java, but also all of the ATL's libraries. Ryan will be able to tell you if any of them are required or not though.

**Comment by dgelessus** [comment]
FWIW, here is a list of libraries which are on the launcher classpath and not included in Forge's libraries:

* ATLauncher classes (obviously)
* json-simple
* jsr305
* IntelliJ stuff (obviously)
* That XZ library we removed

Those are the only libraries that could potentially be a problem. However Forge doesn't need them it seems, and it's unlikely that any mods will rely on ATLauncher-specific behavior anyway. There are enough users who use MultiMC, FTB/Curse or the standard launcher.

**Comment by dgelessus** [comment]
Also, this fix can *not* be applied to `LegacyMCLauncher`, because that class is in the launcher code and passed as the main class to the `java` command. This means that the launcher jar needs to be on the classpath when launching a legacy instance, and `LegacyMCLauncher` already handles this well and doesn't add the entire launcher classpath to Minecraft's unless it cannot find the location of the launcher jar.

**Comment by jamierocks** [comment]
I had wrote some code a while back that made the legacy MC launcher use LaunchWrapper - may turn that into a pr at some point.

**Comment by dgelessus** [comment]
Now also includes a fix for a NPE in `Instance.clone` (https://forums.atlauncher.com/thread-14139.html). I pushed this here rather than in a separate PR because I can't figure out how to branch off the upstream develop without messing up my Git history.

**Pull request closed without merging**